Decrypt KLAP data from PCAP files

clstrickland opened the pull request:
A tool to allow KLAP data to be exported, in JSON, from a PCAP file of encrypted requests.

clstrickland commented:
It looks like it's only unhappy about pyshark being untyped. I would think that, especially since this is just a dev tool, this is something that can safely be ignored?

Aside from that, I think I've convinced Ruff that the code is well-formatted, so it should be good to go.

rytilahti commented on devtools/klap-decrypt-requirements.txt:
```diff
@@ -0,0 +1,61 @@
+aiohttp==3.9.5
```
Instead of pinning the versions, please just add an entry to `README.md` and list direct dependencies there. I think installing `pyshark` is enough?

rytilahti replied:
Don't forget to remove this file.

rytilahti commented on devtools/klap-decrypt.py:
```diff
@@ -0,0 +1,272 @@
+from kasa.klaptransport import KlapEncryptionSession, KlapTransportV2
+
+# These are the only variables that need to be changed in order to run this code.
+# ********** CHANGE THESE **********
+username = "hello@world.com"  # cloud account username (likely an email)
+password = "hunter2"  # cloud account password  # noqa: S105
+# the ip of the smart device as it appears in the pcap file
+device_ip = "192.168.1.100"
+pcap_file_path = "/path/to/my.pcap"  # the path to the pcap file
+output_json_name = (
+    "output.json"  # the name of the output file, relative to the current directory
+)
+# **********************************
```
Please a main entrypoint as function (e.g., `cli()`, see https://github.com/python-kasa/python-kasa/blob/master/devtools/dump_devinfo.py#L211) to allow defining the options from command line.

clstrickland replied:
Done in 4ed43e113b0a488c845c90f8068946512719d4e6

rytilahti commented on devtools/klap-decrypt.py:
```diff
@@ -0,0 +1,272 @@
+)
+
+
+# In case any modifications need to be made in the future,
+# this class is created to allow for easy modification
+class MyKlapTransport(KlapTransportV2):
+    """A custom KlapTransportV2 class that allows for easy modification."""
+
+    pass
```
```suggestion
```
I would skip this for now. We can adapt when and if necessary :-)

clstrickland replied:
Yeah, I don't remember why I thought that was a good idea lol. I think it was an artifact of my original revision, at I for some reason just went with it.

clstrickland replied:
Fixed manually in 83e512fb65df123f42be3e39557d12cb9ccc338c

rytilahti commented on devtools/klap-decrypt.py:
```diff
@@ -0,0 +1,272 @@
+    except StopIteration:
+        break
+
+# save the final array to a file
+with open(output_json_name, "w") as f:
+    f.write(json.dumps(finalArray, indent=2))
+    f.write("\n" * 1)
+    f.close()
```
The command could take `-o` (and `--output`) as an option, and default to printing out the results in the stdout.

clstrickland replied:
Done in fd2dfa12a8552623cf4bba86be6ddef1d5c8fe28

rytilahti reviewed (dismissed):
This is really cool, thanks for working on it! I don't currently have any pcaps to test, but I added some comments to make the script easier to use. On using click, you can take a look into `parse_pcap.py` and `dump_devinfo.py` for some inspiration.

sdb9696 commented on devtools/klap-decrypt.py:
```diff
@@ -0,0 +1,272 @@
+    """A class that handles the data decryption, and the encryption session updating."""
+
+    def __init__(self, klap, creds):
+        self.__local_seed: bytes = None
+        self.__remote_seed: bytes = None
+        self.__session: MyEncryptionSession = None
+        self.__creds = creds
+        self.__klap: MyKlapTransport = klap
+        self.__auth_hash = self.__klap.generate_auth_hash(self.__creds)
+        self.__local_auth_hash = None
+        self.__remote_auth_hash = None
+        self.__seq = 0
```
I think these should be single underscores for consistency with the library

clstrickland replied:
Fixed in 83e512fb65df123f42be3e39557d12cb9ccc338c

sdb9696 reviewed (changes requested):
This looks great, I've ordered a wifi dongle with monitor mode so I can generate a PCAP file to play with. @rytilahti maybe we should add some of the devtools dependencies as a package extra? Or maybe even devtools itself as a package that can be installed as an extra?

Minor change requested to internal variables in additions to @rytilahti comments

clstrickland commented:
@sdb9696 if you have Android you can use PCAPdroid. That's what I used. 

There is an issue with TP-Link's app where, when using PCAPdroid, it thinks it no longer has local access to smart devices. The trick is to open the TP-Link app, establish a connection to the smart device, then start up PCAPdroid. Finally, you need to physically restart the smart device so that the existing handshake is invalidated and it goes through it again now that the packets are being captured.

sdb9696 commented on devtools/klap-decrypt.py:
Can we rename this to `parse_pcap_klap.py` please?

clstrickland replied:
Changed with 4f56c35655e2800e849a2637f551af75b15540ed

clstrickland commented:
> I didn't get any issues with odd characters and I'm wondering if it's because you're getting the sequence numbers from all the devices rather than the just the `device_ip` you want.

I was able to get that working before making the PR. I had tried using seq numbers to no avail when I made that original comment about the weird characters. It turns out that, in my original script, I was setting the seq number but then it was being overridden by internal logic for the encryption. Once I fixed that, it worked perfectly.

sdb9696 commented:
> I was able to get that working before making the PR. I had tried using seq numbers to no avail when I made that original comment about the weird characters. It turns out that, in my original script, I was setting the seq number but then it was being overridden by internal logic for the encryption. Once I fixed that, it worked perfectly.

Cool, I would still move the:
```
 if packet.ip.dst != device_ip:
                        continue
``` 
up above the sequence number bit though no?

Are you good with all the review comments? It'd be nice to get this merged soonish as we could ask a few people to run it against some of their more interesting devices once the `cli` parameters are added.

clstrickland commented:
> > I was able to get that working before making the PR. I had tried using seq numbers to no avail when I made that original comment about the weird characters. It turns out that, in my original script, I was setting the seq number but then it was being overridden by internal logic for the encryption. Once I fixed that, it worked perfectly.
> 
> Cool, I would still move the:
> 
> ```
>  if packet.ip.dst != device_ip:
>                         continue
> ```
> 
> up above the sequence number bit though no?

I could not put it any higher up because in handshake1 we need to grab the response, which will not have a matching device IP. If I put it any further above, it would strip out that packet completely.

> Are you good with all the review comments? It'd be nice to get this merged soonish as we could ask a few people to run it against some of their more interesting devices once the `cli` parameters are added.

I'm going to see what I can do right now, but my power (and therefore internet connection) has fallen victim to Hurricane Beryl, so I'm kinda at the mercy of Mother Nature right now 🥲

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+        self._local_auth_hash = None
+        self._remote_auth_hash = None
+        self._seq = 0
+        pass
```
```suggestion
```

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+        device_ip, connection_type=fakeConnection, credentials=myCreds
+    )
+
+
+
```
```suggestion
```

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+
+    operator = Operator(KlapTransportV2(config=fakeDevice), myCreds)
+
+    finalArray = []
```
```suggestion
    packets = []
```

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+
+
+if __name__ == "__main__":
+    cli()
```
If you run `pre-commit run -a`, it should fix issues like this :-)

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+                        message = bytes.fromhex(data.replace(":", ""))
+                        try:
+                            plaintext = operator.decrypt(message)
+                            myDict = json.loads(plaintext)
```
```suggestion
                            payload = json.loads(plaintext)
```

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+    # In an effort to keep this code tied into the original code
+    # (so that this can hopefully leverage any future codebase updates inheriently),
+    # some weird initialization is done here
+    myCreds = Credentials(username, password)
```
```suggestion
    creds = Credentials(username, password)
```

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+    # some weird initialization is done here
+    myCreds = Credentials(username, password)
+
+    fakeConnection = DeviceConnectionParameters(
+        DeviceFamily.SmartTapoBulb, DeviceEncryptionType.Klap
+    )
+    fakeDevice = DeviceConfig(
+        device_ip, connection_type=fakeConnection, credentials=myCreds
+    )
```
```suggestion
    fake_connection = DeviceConnectionParameters(
        DeviceFamily.SmartTapoBulb, DeviceEncryptionType.Klap
    )
    fake_device = DeviceConfig(
        device_ip, connection_type=fakeConnection, credentials=myCreds
    )
```
Use camel_case for variable naming, CamelCase is only reserved for class names & enum members.

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+        if self._session is None:
+            raise ValueError("No session available")
+        return self._session.decrypt(*args, **kwargs)
+
+
```
```suggestion
```

rytilahti commented on devtools/parse_pcap_klap.py:
```diff
@@ -0,0 +1,300 @@
+
+
+codecs.register_error("bad_chars_replacement", bad_chars_replacement)
+
+
+
+
```
```suggestion
```

rytilahti reviewed (commented):
Some more comments, this looks pretty much ready to go, thanks @clstrickland!

clstrickland commented:
@rytilahti I renamed symbols. I just used the pre-commit command, so I assume it automatically fixed all those spacing issues. I think I stopped running it at some point because (part of) it always failed due to pyshark being untyped.

As far as dependencies go, I agree with @sdb9696 (https://github.com/python-kasa/python-kasa/pull/1041#pullrequestreview-2171762463) that it should somehow be included in the existing package management solution. I removed the requirements file and put a note in the readme, but it would still be really cool to see it in a devtools extra.

sdb9696 reviewed (approved):
LGTM, fixed the CI and made some tweaks to get it working on linux and with `asyncclick`

Thanks for the great PR @clstrickland !
